README.md
# baseplate-utils


CHANGELOG.md
# @baseplate-sdk/utils

## 3.2.1

### Patch Changes

- [#13](https://github.com/ConvexCooperative/baseplate-utils/pull/13) [`7b3303a`](https://github.com/ConvexCooperative/baseplate-utils/commit/7b3303a4c07bd3302019cbc738fca839d3699e4c) Thanks [@joeldenning](https://github.com/joeldenning)! - Add isProd property to CustomDomain type

## 3.2.0

### Minor Changes

- [#12](https://github.com/ConvexCooperative/baseplate-utils/pull/12) [`76faf11`](https://github.com/ConvexCooperative/baseplate-utils/commit/76faf1111a2b261326636ace4ff982a717b13f7e) Thanks [@joeldenning](https://github.com/joeldenning)! - Implement mergeDefaultHtmlParams function

### Patch Changes

- [#12](https://github.com/ConvexCooperative/baseplate-utils/pull/12) [`76faf11`](https://github.com/ConvexCooperative/baseplate-utils/commit/76faf1111a2b261326636ace4ff982a717b13f7e) Thanks [@joeldenning](https://github.com/joeldenning)! - Add typings for CustomDomain values in KV storage

## 3.1.3

### Patch Changes

- [#10](https://github.com/ConvexCooperative/baseplate-utils/pull/10) [`702c359`](https://github.com/ConvexCooperative/baseplate-utils/commit/702c3597f7d153dd72bfd7ec68fa0d4346033d10) Thanks [@asekito](https://github.com/asekito)! - Custom domain settings for OrgSetting type

## 3.1.2

### Patch Changes

- [#9](https://github.com/ConvexCooperative/baseplate-utils/pull/9) [`4b2571f`](https://github.com/ConvexCooperative/baseplate-utils/commit/4b2571f0b4899d3d56d38a01da7df7ed46246113) Thanks [@joeldenning](https://github.com/joeldenning)! - Fix handling of null values

## 3.1.1

### Patch Changes

- [#8](https://github.com/ConvexCooperative/baseplate-utils/pull/8) [`a703951`](https://github.com/ConvexCooperative/baseplate-utils/commit/a703951fce06470fc60914bcddf774ba4c55af7d) Thanks [@joeldenning](https://github.com/joeldenning)! - Add environmentId to proxysettings

## 3.1.0

### Minor Changes

- [#7](https://github.com/ConvexCooperative/baseplate-utils/pull/7) [`254662d`](https://github.com/ConvexCooperative/baseplate-utils/commit/254662d091ee5cb3e9e1f98277ddea098416a0b4) Thanks [@Ksouffle](https://github.com/Ksouffle)! - added aws secrets to StaticFileProxySettings

## 3.0.0

### Major Changes

- [#6](https://github.com/ConvexCooperative/baseplate-utils/pull/6) [`e9a9765`](https://github.com/ConvexCooperative/baseplate-utils/commit/e9a9765620bf474639cee323f8c1b421bc39da93) Thanks [@joeldenning](https://github.com/joeldenning)! - Rename customHost to host

## 2.0.0

### Major Changes

- [#5](https://github.com/ConvexCooperative/baseplate-utils/pull/5) [`837dd62`](https://github.com/ConvexCooperative/baseplate-utils/commit/837dd62eac8883ca48661d767c1855629a9fced8) Thanks [@joeldenning](https://github.com/joeldenning)! - Switch npm scope and license. Now published under `@baseplate-sdk/utils`. License is now AGPL 3 + Commons Clause.

## 1.0.3

### Patch Changes

- [#4](https://github.com/ConvexCooperative/baseplate-utils/pull/4) [`879e819`](https://github.com/ConvexCooperative/baseplate-utils/commit/879e81975b2c32dc911121691f1c84f4eab03ef4) Thanks [@Ksouffle](https://github.com/Ksouffle)! - Added cache-control header to default settings

## 1.0.2

### Patch Changes

- [#3](https://github.com/ConvexCooperative/baseplate-utils/pull/3) [`d57383e`](https://github.com/ConvexCooperative/baseplate-utils/commit/d57383eb5a6333e1725ffa68952d57bf734ec347) Thanks [@joeldenning](https://github.com/joeldenning)! - Fix file extension weird things

## 1.0.1

### Patch Changes

- [#2](https://github.com/ConvexCooperative/baseplate-utils/pull/2) [`e3fba73`](https://github.com/ConvexCooperative/baseplate-utils/commit/e3fba73067efa123ffcbf4737c8e627dc7eee213) Thanks [@joeldenning](https://github.com/joeldenning)! - Add package.json "main" field for old webpack


src/htmlParams.test.ts
import { HTMLTemplateParams, mergeDefaultHtmlParams } from "./htmlParams.js";
import { RecursivePartial } from "./utils.js";

describe(`htmlParams`, () => {
  it(`matches snapshot for default params`, () => {
    const params = mergeDefaultHtmlParams({});
    expect(params).toMatchSnapshot();
  });

  it(`merges top level properties`, () => {
    const params: RecursivePartial<HTMLTemplateParams> = {
      lang: "es",
      pageTitle: "Aplicacion de Web",
      preloads: [
        {
          importSpecifier: "root-config",
          as: "script",
        },
      ],
    };
    const finalParams = mergeDefaultHtmlParams(params);
    expect(finalParams.lang).toBe(params.lang);
    expect(finalParams.pageTitle).toBe(params.pageTitle);
    expect(finalParams.preloads).toEqual(params.preloads);
  });

  it(`merges import map properties`, () => {
    const params: RecursivePartial<HTMLTemplateParams> = {
      importMap: {
        name: "main",
        useOverrides: false,
      },
    };
    const finalParams = mergeDefaultHtmlParams(params);
    expect(finalParams.importMap.type).toBe("systemjs");
    expect(finalParams.importMap.name).toBe(params.importMap!.name);
    expect(finalParams.importMap.useOverrides).toBe(
      params.importMap!.useOverrides
    );
  });
});


src/orgSettings.test.ts
import { mergeDefaultOrgSettings } from "./orgSettings.js";

describe(`orgSettings`, () => {
  describe(`mergeDefaultOrgSettings`, () => {
    it(`sets a default import map cache control`, () => {
      const finalSettings = mergeDefaultOrgSettings({});

      expect(finalSettings.importMapCacheControl).toBe(
        "public, must-revalidate, max-age=60"
      );
    });

    it(`sets a default cache control for static files`, () => {
      const finalSettings = mergeDefaultOrgSettings({});

      expect(finalSettings.staticFiles.cacheControl).toBe(
        "public, max-age=31536000, immutable"
      );
    });

    it(`overwrites null values with the default settings`, () => {
      const finalSettings = mergeDefaultOrgSettings({
        cors: {
          allowCredentials: null,
          allowHeaders: null,
          allowMethods: null,
          allowOrigins: null,
          exposeHeaders: null,
          maxAge: null,
        },
        customDomain: {
          customCDNTestDomain: null,
          customCDNProdDomain: null,
          customCDNTestCloudflareIdentifier: null,
          customCDNProdCloudflareIdentifier: null,
        },
      });

      expect(finalSettings.cors.allowCredentials).toEqual(true);
      expect(finalSettings.cors.allowOrigins).toEqual(["*"]);
      expect(finalSettings.cors.exposeHeaders).toEqual([]);
      expect(finalSettings.cors.maxAge).toEqual(86400);
      expect(finalSettings.cors.allowCredentials).toEqual(true);
      expect(finalSettings.cors.allowHeaders).toEqual([]);
      expect(finalSettings.cors.allowMethods).toEqual([
        "GET",
        "HEAD",
        "OPTIONS",
      ]);
      expect(finalSettings.customDomain.customCDNTestDomain).toBeNull();
      expect(
        finalSettings.customDomain.customCDNTestCloudflareIdentifier
      ).toBeNull();
      expect(finalSettings.customDomain.customCDNProdDomain).toBeNull();
      expect(
        finalSettings.customDomain.customCDNProdCloudflareIdentifier
      ).toBeNull();
    });
  });
});


src/htmlParams.ts
import { mergeWith } from "lodash-es";
import { RecursivePartial } from "./utils.js";

const defaultParams: HTMLTemplateParams = {
  lang: "en",
  pageTitle: "Web App",
  importMap: {
    type: "systemjs",
    name: "systemjs",
    useOverrides: true,
  },
  pageInit: {
    type: "single-spa",
    layoutTemplate: "<single-spa-router></single-spa-router>",
  },
  preloads: [],
  headScripts: [],
  headStylesheets: [],
};

export function mergeDefaultHtmlParams(
  htmlParams: RecursivePartial<HTMLTemplateParams>
): HTMLTemplateParams {
  const finalSettings: HTMLTemplateParams = mergeWith(
    {},
    defaultParams,
    htmlParams,
    (newValue, oldValue) => {
      return oldValue === null ? newValue : undefined;
    }
  );

  return finalSettings;
}

export interface HTMLTemplateParams {
  lang: string;
  pageTitle: string;
  importMap: {
    type: "systemjs" | "native";
    name: string;
    useOverrides: boolean;
  };
  pageInit: HTMLPageInit;
  preloads: HTMLPreload[];
  headScripts: HTMLScript[];
  headStylesheets: HTMLStylesheet[];
}

export type HTMLPageInit = ModulePageInit | SingleSpaPageInit;

export interface ModulePageInit {
  type: "module";
  entryModule: string;
}

export interface SingleSpaPageInit {
  type: "single-spa";
  layoutTemplate: string;
}

export interface HTMLPreload {
  importSpecifier?: string;
  href?: string;
  as: string;
}

export interface HTMLScript {
  src: string;
  async: boolean;
  defer: boolean;
}

export interface HTMLStylesheet {
  href: string;
}


src/orgSettings.ts
import { mergeWith } from "lodash-es";
import { RecursivePartial } from "./utils.js";

const defaultSettings: OrgSettings = {
  staticFiles: {
    microfrontendProxy: {
      environments: {},
    },
    cacheControl: "public, max-age=31536000, immutable",
  },
  importMapCacheControl: "public, must-revalidate, max-age=60",
  cors: {
    allowOrigins: ["*"],
    exposeHeaders: [],
    // 1 day in seconds
    maxAge: 86400,
    allowCredentials: true,
    allowHeaders: [],
    allowMethods: ["GET", "HEAD", "OPTIONS"],
  },
  orgExists: false,
  customDomain: {
    customCDNTestDomain: null,
    customCDNProdDomain: null,
    customCDNTestCloudflareIdentifier: null,
    customCDNProdCloudflareIdentifier: null,
  },
};

export function mergeDefaultOrgSettings(
  orgSettings: RecursivePartial<OrgSettings>
): OrgSettings {
  const finalSettings: OrgSettings = mergeWith(
    {},
    defaultSettings,
    orgSettings,
    (newValue, oldValue) => {
      return oldValue === null ? newValue : undefined;
    }
  );
  return finalSettings;
}

interface StaticFileSettings {
  microfrontendProxy: {
    environments: {
      [key: string]: StaticFileProxySettings;
    };
  };
  cacheControl: string;
}

export interface StaticFileProxySettings {
  useBaseplateHosting: boolean;
  host: string;
  environmentId: string;
  aws?: {
    region: string;
    accessKeyId: string;
    secretAccessKey: string;
  };
}

export interface OrgSettings {
  staticFiles: StaticFileSettings;
  importMapCacheControl: string;
  cors: CORSSettings;
  orgExists: boolean;
  customDomain: CustomDomainSettings;
}

export interface CORSSettings {
  allowOrigins: string[];
  exposeHeaders: string[];
  maxAge: number;
  allowCredentials: boolean;
  allowMethods: string[];
  allowHeaders: string[];
}

export interface CustomDomainSettings {
  customCDNTestDomain?: string;
  customCDNProdDomain?: string;
  customCDNTestCloudflareIdentifier?: string;
  customCDNProdCloudflareIdentifier?: string;
}
